### PR TITLE
adds monkeys to the game

### LINKED ISSUE #11518: Monkeys cant cryo
### Severity

Inconvenience

### Short Description

Monkified humans cant cryo

### Reproduction

1. get monkified via a monkey genetic activator/mutator
2. try to go to cryo via a cryogenic freezer

### Round ID

Round #35465

### Test Merges

#11366#11425#11413#11139#11464#11378

### BYOND Username

gboster

### Additional information

This is just painfull as you have to go to a genetecist/chemist if there is one and get mutadone

### LINKED ISSUE #13187: Being able to bite as monkey thru walls and from far away
### Severity

Exploit

### Short Description

Found myself being able to bite a pearson thru a window

### Reproduction

**1**.Join round as normal crew member ( i was a lizard).
**2**.Get monkefied by using a mutator from genetics.
**3**.Cuff yourself, works with any type of cuffs
**4**.Harm click on people.

### Round ID

39318

### Test Merges

#13033 #12948 #12912 #12959

### BYOND Username

Tosterimc

### Additional information

If the steps wont work it may been coused couse of the countless stuning and other stuff that happend to me before i found out.
*edit: this only works when being cuffed as the monkey otherwise it wont work.

### LINKED ISSUE #17618: Nuggeted monkeys render with limbs
### Severity

Inconvenience

### Short Description

turn monke into nugget and you can still see limbs

### Reproduction

1. spawn monke
2. saw its limbs off with BRUTE FORCE
3. why do i still see limbs

### Round ID

45044

### Test Merges

_No response_

### BYOND Username

Addust

### Was this done on a locally hosted or non-Yogstation hosted server?

No

### Additional information

oog

### LINKED ISSUE #18132: Monkey has Polymorph legs when changing
## Test Merges: 
Not Supplied / None
## Reproduction:
Details: was human, turned into polymorph, then decided to be monkey and accidently created this monstrocity
## Supplied Image:

SuppliedImage
 Submitted by: EnglishScotsman (740223180806946946)

### LINKED ISSUE #18604: Monkeys can steal bluespace bags from inside of bluespace bags
## Test Merges: 
Not Supplied / None
## Reproduction:
Monkeys are able to steal bluespace bags from you while inside of them, causing the bag to vanish into the ether. Eventually the monkey inside of the bag will drop the bag, causing the bag to reappear in your inventory glitched.
## Supplied Image:

SuppliedImage
 Submitted by: Zer0 (412038981669355560)

### PR DESCRIPTION
**Pull request by tattax**
PR name is just to keep suit and tell any recurring viewers what is gonna happen here:

tg ports,
cause it turns out to port mob AI we need a bajillion other things

Closes: https://github.com/yogstation13/Yogstation/issues/18604
Closes: https://github.com/yogstation13/Yogstation/issues/11518
Closes: https://github.com/yogstation13/Yogstation/issues/18132
Closes: https://github.com/yogstation13/Yogstation/issues/17618
Closes: https://github.com/yogstation13/Yogstation/issues/13187

# Document the changes in your pull request

ports movement subsystem from TG:
https://github.com/tgstation/tgstation/pull/62567
https://github.com/tgstation/tgstation/pull/66628
https://github.com/tgstation/tgstation/pull/66567

JPS pathfinding system:
https://github.com/tgstation/tgstation/pull/56780
https://github.com/tgstation/tgstation/pull/70623

and most importantly (the goal of the pr) mob AI:
https://github.com/tgstation/tgstation/pull/55238
https://github.com/tgstation/tgstation/pull/61726

and monkey things:
https://github.com/tgstation/tgstation/pull/66057
https://github.com/tgstation/tgstation/pull/68489
https://github.com/tgstation/tgstation/pull/55614

this was originally gonna be a mice pr where i add some mice and we all have fun but we can't have that because we don't have the infrastructure. This is also why you'll see some mice sprites in icon diff bot (htey don't add anything) (yet.)

qDr8w5T7db

aGaemVIMbs

image
^ why this needs to be merged

:cl: Qustinnus, LemonInTheDark, Ryll-Ryll
rscadd: mobs that use mob AI are better at pathfinding and have more interactions
rscadd: space drifting is smoother
bugfix: fixes some monkey bugs (see above)
soundadd: new chimp sounds
/:cl:

### CODE REVIEW

**Comment by BurgerLUA** [comment]
pretty based pr but good luck reviewing everything holy shit

**Comment by YashinHak** [comment]
Based. Chimpilled.

**Review by cuackles — CHANGES_REQUESTED** [review]
it deletes a head worn icon, pls fix

File: code/__HELPERS/type2type.dm [hunk 1]
```diff
@@ -358,14 +358,14 @@
...
-/proc/body_parts_covered2organ_names(bpc)
+//(I challenge you to find a use for this) -I found a use for it!! | So did I!.
+/proc/cover_flags2body_zones(bpc)
 	var/list/covered_parts = list()
 
 	if(!bpc)
 		return 0
 
-	if(bpc & FULL_BODY)
+	if(bpc == FULL_BODY)
```
Comment (BurgerLUA) [inline code/__HELPERS/type2type.dm:367]: I think this change should be reverted as it will cover more cases.

Comment (tattax) [reply code/__HELPERS/type2type.dm:367]: i could, but TG does this way and i'm sure it's supposed to be more precise

**Comment by MajManatee** [comment]
monke

**Review by cuackles — CHANGES_REQUESTED** [review]
can you put the old rat king spirte back ? the new one seems a bit to over the top in comparation with all the other things

File: code/__DEFINES/DNA.dm [hunk 2]
```diff
@@ -155,6 +140,10 @@
 /// has a tail
 #define HAS_TAIL		28
 #define NONANITES		29
+/// Do not draw blood overlay
+#define NOBLOODOVERLAY 	30
+///No augments, for monkeys in specific because they will turn into fucking freakazoids https://cdn.discordapp.com/attachments/326831214667235328/791313258912153640/102707682-fa7cad80-4294-11eb-8f13-8c689468aeb0.png
+#define NOAUGMENTS 		31
```
Comment (JohnFulpWillard) [inline code/__DEFINES/DNA.dm]:
```suggestion
#define NOAUGMENTS 31
```

File: code/__DEFINES/DNA.dm [hunk 3]
```diff
@@ -155,6 +140,10 @@
 /// has a tail
 #define HAS_TAIL		28
 #define NONANITES		29
+/// Do not draw blood overlay
+#define NOBLOODOVERLAY 	30
```
Comment (JohnFulpWillard) [inline code/__DEFINES/DNA.dm]:
```suggestion
#define NOBLOODOVERLAY 30
```

File: code/__DEFINES/DNA.dm [hunk 4]
```diff
@@ -143,9 +128,9 @@
 #define NOEYESPRITES	22
 /// Used for determining which wounds are applicable to this species.
 /// if we have flesh (can suffer slash/piercing/burn wounds, requires they don't have NOBLOOD)
-#define HAS_FLESH	23
+#define HAS_FLESH		23
 /// if we have bones (can suffer bone wounds)
-#define HAS_BONE	24
+#define HAS_BONE		24
```
Comment (JohnFulpWillard) [inline code/__DEFINES/DNA.dm]:
```suggestion
#define HAS_BONE 24
```

File: code/__DEFINES/DNA.dm [hunk 5]
```diff
@@ -143,9 +128,9 @@
 #define NOEYESPRITES	22
 /// Used for determining which wounds are applicable to this species.
 /// if we have flesh (can suffer slash/piercing/burn wounds, requires they don't have NOBLOOD)
-#define HAS_FLESH	23
+#define HAS_FLESH		23
```
Comment (JohnFulpWillard) [inline code/__DEFINES/DNA.dm]:
```suggestion
#define HAS_FLESH 23
```

File: code/__DEFINES/DNA.dm [hunk 6]
```diff
@@ -155,6 +140,10 @@
 /// has a tail
 #define HAS_TAIL		28
 #define NONANITES		29
+/// Do not draw blood overlay
+#define NOBLOODOVERLAY 	30
+///No augments, for monkeys in specific because they will turn into fucking freakazoids https://cdn.discordapp.com/attachments/326831214667235328/791313258912153640/102707682-fa7cad80-4294-11eb-8f13-8c689468aeb0.png
```
Comment (JohnFulpWillard) [inline code/__DEFINES/DNA.dm:173]: Discordapp images are not a good idea if you want this future-proofed.

File: code/__DEFINES/food.dm [hunk 7]
```diff
@@ -44,3 +44,6 @@
 #define DRINK_VERYGOOD	3
 #define DRINK_FANTASTIC	4
 #define FOOD_AMAZING 5
+
+//#define IS_EDIBLE(O) (O.GetComponent(/datum/component/edible))
```
Comment (JohnFulpWillard) [inline code/__DEFINES/food.dm:49]: This comment seems useless

File: code/__DEFINES/is_helpers.dm [hunk 8]
```diff
@@ -89,9 +89,7 @@ GLOBAL_LIST_INIT(turfs_without_ground, typecacheof(list(
 #define issnail(A) (is_species(A, /datum/species/snail))
 #define isandroid(A) (is_species(A, /datum/species/android))
 #define isdummy(A) (istype(A, /mob/living/carbon/human/dummy))
-
-//more carbon mobs
-#define ismonkey(A) (istype(A, /mob/living/carbon/monkey))
+#define ismonkey(A) (is_species(A, /datum/species/monkey))
```
Comment (JohnFulpWillard) [inline code/__DEFINES/is_helpers.dm:99]: Move this to the species is_helpers

File: code/__DEFINES/mob_ai.dm [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+// hostile command modes (what pointing at something/someone does depending on the last order the carp heard)
+/// Don't do anything (will still react to stuff around them though)
+#define HOSTILE_COMMAND_NONE 0
+/// Will attack a target.
+#define HOSTILE_COMMAND_ATTACK 1
+/// Will follow a target.
+#define HOSTILE_COMMAND_FOLLOW 2
+
+///Dog AI controller blackboard keys
```
Comment (JohnFulpWillard) [inline code/__DEFINES/mob_ai.dm:169]: We don't need Dog AI do we?

File: code/__DEFINES/mob_ai.dm [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+///Baby-making blackboard
+///Types of animal we can make babies with.
+#define BB_BABIES_PARTNER_TYPES "BB_babies_partner"
+///Types of animal that we make as a baby.
+#define BB_BABIES_CHILD_TYPES "BB_babies_child"
+///Current partner target
+#define BB_BABIES_TARGET "BB_babies_target"
+
+// Bileworm AI keys
```
Comment (JohnFulpWillard) [inline code/__DEFINES/mob_ai.dm:258]: or really anything else in here except monkey

File: code/__DEFINES/time.dm [hunk 11]
```diff
@@ -21,6 +21,7 @@
 #define HALLOWEEN				"Halloween"
 #define CHRISTMAS				"Christmas"
 #define FESTIVE_SEASON			"Festive Season"
+#define MONKEYDAY				 "Monkey Day"
```
Comment (JohnFulpWillard) [inline code/__DEFINES/time.dm]:
```suggestion
#define MONKEYDAY "Monkey Day"
```

File: code/__HELPERS/mob_ai.dm [hunk 12]
```diff
@@ -0,0 +1,47 @@
+//this file has procs that help ais think good. used by behaviors mostly
+
+
+/// Returns either the best weapon from the given choices or null if held weapons are better
+/proc/GetBestWeapon(datum/ai_controller/controller, list/choices, list/held_weapons)
```
Comment (JohnFulpWillard) [inline code/__HELPERS/mob_ai.dm]: snake case please

File: code/__HELPERS/mob_ai.dm [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+			// Also yes it is intentional that pawns dont know how to pick the best gun
+			return item
+		if(item.force > top_force)
+			top_force = item.force
+			top_force_item = item
+
+	for(var/obj/item/item as anything in choices)
+		if(!item)
+			continue
+//		if(HAS_TRAIT(item, TRAIT_NEEDS_TWO_HANDS) || controller.blackboard[BB_MONKEY_BLACKLISTITEMS][item])
```
Comment (JohnFulpWillard) [inline code/__HELPERS/mob_ai.dm]: Useless comment, please remove

File: code/__HELPERS/mob_ai.dm [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+			return item
+		if(item.force <= top_force)
+			continue
+		top_force_item = item
+		top_force = item.force
+
+	return top_force_item
+
+///returns if something can be consumed, drink or food
+/proc/IsEdible(obj/item/thing)
```
Comment (JohnFulpWillard) [inline code/__HELPERS/mob_ai.dm:37]: You have the IS_EDIBLE define already, this is useless, drinking glasses return TRUE by that.

File: code/__HELPERS/mobs.dm [hunk 15]
```diff
@@ -118,7 +118,8 @@
 		"pod_hair" = pick(GLOB.pod_hair_list),
 		"ipc_screen" = pick(GLOB.ipc_screens_list),
 		"ipc_antenna" = pick(GLOB.ipc_antennas_list),
-		"ipc_chassis" = pick(GLOB.ipc_chassis_list)
+		"ipc_chassis" = pick(GLOB.ipc_chassis_list),
+		"tail_monkey" = "None"
```
Comment (JohnFulpWillard) [inline code/__HELPERS/mobs.dm:130]: If you're already making a list of monkey tails, why not make it pick one of them here?

File: code/_onclick/hud/parallax.dm [hunk 16]
```diff
@@ -326,12 +347,30 @@
...
+		COMSIG_MOB_LOGOUT = PROC_REF(on_mob_logout),
+	)
+	AddComponent(/datum/component/connect_mob_behalf, owner.client, connections)
+	on_z_change(owner)
+
+/atom/movable/screen/parallax_layer/planet/proc/on_mob_logout(mob/source)
+	SIGNAL_HANDLER
+//	var/client/boss = source.canon_client
+	var/client/boss = source.client
+	on_z_change(boss.mob)
```
Comment (JohnFulpWillard) [inline code/_onclick/hud/parallax.dm]: Why not just use ``source`` as the arg? Does a mob even have a client when they've logged out?

File: code/_onclick/other_mobs.dm [hunk 17]
```diff
@@ -135,7 +133,7 @@
 	moving it here instead of various hand_p's has simplified
 	things considerably
 */
-/mob/living/carbon/monkey/RestrainedClickOn(atom/A)
+/mob/living/carbon/human/species/monkey/RestrainedClickOn(atom/A)
```
Comment (JohnFulpWillard) [inline code/_onclick/other_mobs.dm]: Shouldn't this be on the species?

File: code/datums/components/jetpack.dm [hunk 18]
```diff
@@ -0,0 +1,149 @@
+// Welcome to the jetpack component
+// Apply this to something when you want it to be "like a jetpack"
+// So propulsion through space on move, that sort of thing
+/datum/component/jetpack
```
Comment (JohnFulpWillard) [inline code/datums/components/jetpack.dm:4]: Can you mention jetpack changes in the changelog

File: code/datums/components/jetpack.dm [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+/datum/component/jetpack/proc/setup_trail()
+	var/mob/moving = get_mover.Invoke()
+	if(!moving || trail)
+		return
+	trail = new effect_type
+	trail.auto_process = FALSE
+	trail.set_up(moving)
+
+/datum/component/jetpack/proc/activate(datum/source)
+//	SIGNAL_HANDLER
```
Comment (JohnFulpWillard) [inline code/datums/components/jetpack.dm]: Why no Signal handlers?

File: code/game/mecha/mecha.dm [hunk 20]
```diff
@@ -624,21 +624,30 @@
...
-			if(backup.newtonian_move(turn(movement_dir, 180)))
-				if(occupant)
-					to_chat(occupant, span_info("You push off of [backup] to propel yourself."))
+		if(!istype(backup) || !movement_dir || backup.anchored || continuous_move) //get_spacemove_backup() already checks if a returned turf is solid, so we can just go
+			return TRUE
+		last_pushoff = world.time
+		if(backup.newtonian_move(turn(movement_dir, 180), instant = TRUE))
+			backup.last_pushoff = world.time
+//			step_silent = TRUE
+//			if(return_drivers())
```
Comment (JohnFulpWillard) [inline code/game/mecha/mecha.dm:587]: :trollface:

File: code/game/objects/effects/decals/cleanable/aliens.dm [hunk 21]
```diff
@@ -26,15 +26,26 @@
...
+				break
+		return
+
+	var/datum/move_loop/loop = SSmove_manager.move_to_dir(src, get_step(src, direction), delay = delay, timeout = range * delay, priority = MOVEMENT_ABOVE_SPACE_PRIORITY)
+	RegisterSignal(loop, COMSIG_MOVELOOP_POSTPROCESS, PROC_REF(spread_movement_effects))
+
+
+/obj/effect/decal/cleanable/xenoblood/xgibs/proc/spread_movement_effects(datum/move_loop/has_target/source)
+	SIGNAL_HANDLER
+	new /obj/effect/decal/cleanable/xenoblood/xsplatter(loc)
```
Comment (JohnFulpWillard) [inline code/game/objects/effects/decals/cleanable/aliens.dm]: Does this work?
It seems really odd.

**Review by JohnFulpWillard — COMMENTED** [review]
You're missing an important port - https://github.com/tgstation/tgstation/pull/62125

File: code/__DEFINES/DNA.dm (hunk 6 above)
Comment (tattax) [reply code/__DEFINES/DNA.dm:173]: it is the same as teegee stassion

File: code/__DEFINES/is_helpers.dm (hunk 8 above)
Comment (tattax) [reply code/__DEFINES/is_helpers.dm:99]: i did

File: code/__DEFINES/food.dm (hunk 7 above)
Comment (tattax) [reply code/__DEFINES/food.dm:49]: i am saving it for the future of mankind, it's kind of like cryoing someone irl

File: code/__DEFINES/mob_ai.dm (hunk 9 above)
Comment (tattax) [reply code/__DEFINES/mob_ai.dm:169]: updated file is good for future ports imo

File: code/datums/components/jetpack.dm (hunk 19 above)
Comment (tattax) [reply code/datums/components/jetpack.dm]: they keep fucking up lints and i comment them out to return them up in the near future where yogstation purges all the sleeps and stoplags that are cuasing this issue

**Review by cuackles — APPROVED** [review]

**Review by cuackles — CHANGES_REQUESTED** [review]
image
u did it again

File: code/_onclick/other_mobs.dm (hunk 17 above)
Comment (tattax) [reply code/_onclick/other_mobs.dm]: this actually should be removed

File: code/datums/components/jetpack.dm (hunk 18 above)
Comment (tattax) [reply code/datums/components/jetpack.dm:4]: it functions about the same as it used to

**Comment by tattax** [comment]
relies on the merge of: https://github.com/yogstation13/Yogstation/pull/19165 to fix some issues

**Review by cuackles — APPROVED** [review]

File: code/__DEFINES/is_helpers.dm (hunk 8 above)
Comment (JohnFulpWillard) [reply code/__DEFINES/is_helpers.dm:99]: It's between the check for dummy and alien, they arent friggin species

**Comment by MajManatee** [comment]
hurry up with the monkey already smh

**Comment by MajManatee** [comment]
Reopen when work continues. its been almost 2 months.

**Comment by AMyriad** [comment]
one day...

**Comment by ToasterBiome** [comment]
can this be split into sub PRs? 500 files changed…

**Review by ToasterBiome — CHANGES_REQUESTED** [review]
revert the get_organ_slot change, at the minimum that should be a separate PR

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 22]
```diff
@@ -665,8 +665,8 @@
 		//if they are holding or wearing a card that has access, that works
 		if(check_access(H.get_active_held_item()) || check_access(H.wear_id))
 			return TRUE
-	else if(ismonkey(M))
-		var/mob/living/carbon/monkey/george = M
+	else if(isalien(M))
```
Comment (ToasterBiome) [inline code/modules/mob/living/silicon/robot/robot.dm:677]: why is this changed to alien?

**Review by ToasterBiome — CHANGES_REQUESTED** [review]
revert PASSMACHINE change too.. again, seperate PR

File: code/datums/ai_controllers/_ai_controller.dm [hunk 23]
```diff
@@ -0,0 +1,338 @@
...
+/// Returns what the AI status should be based on current conditions.
+/datum/ai_controller/proc/get_expected_ai_status()
+	var/final_status = AI_STATUS_ON
+
+	if (!ismob(pawn))
+		return final_status
+
+	var/mob/living/mob_pawn = pawn
+
+	if(!continue_processing_when_client && mob_pawn.client)
```
Comment (JohnFulpWillard) [inline code/datums/ai_controllers/_ai_controller.dm:129]: I'm on mobile but make sure you've also got the new update z level mob procs for this

Comment (tattax) [reply code/datums/ai_controllers/_ai_controller.dm:129]: thanks b

**Comment by Moltijoe** [comment]
Reopen this when you actually plan on working on it.
Dont just reopen it, do one change, and then abandon it again

File: code/game/objects/items/devices/miragedrive.dm [hunk 24]
```diff
@@ -14,26 +14,11 @@
...
-/obj/item/mirage_drive/Destroy(force=FALSE)
-	QDEL_NULL(access_card)
-	moving = null
-	return ..()
-
-/obj/item/mirage_drive/examine(datum/source, mob/user, list/examine_list)
-	. = ..()
-	if(!COOLDOWN_FINISHED(src, last_dash))
-		. += span_notice("A digital display on it reads [COOLDOWN_TIMELEFT(src, last_dash)/10].")
-
```
Comment (Moltijoe) [inline code/game/objects/items/devices/miragedrive.dm]: Whats all this about?

Comment (Moltijoe) [reply code/game/objects/items/devices/miragedrive.dm]: I plan to take another look at some point, but i'll see if i can get a testmerge going

Comment (tattax) [reply code/game/objects/items/devices/miragedrive.dm]: auto merge bug, i'll fix

Comment (tattax) [reply code/game/objects/items/devices/miragedrive.dm]: first 2 procs are because you don't need ID card anymore to test for pathing

Comment (Moltijoe) [reply code/game/objects/items/devices/miragedrive.dm]: fair enough, the examine is still needed to figure out the cooldown though

**Comment by tattax** [comment]
won't be able to work on it till july 15

**Comment by ToasterBiome** [comment]
Reopen with more work

**Pull request closed without merging**